ext/drx/drx_time_scale.c

```diff
@@ -0,0 +1,608 @@
+        int id = -1;
+        int found = dr_sscanf(linebuf, "ID: %d", &id);
+        if (found == 1 && id >= 0) {
+#ifdef X64
+            struct itimerspec spec;
+            /* We use dr_invoke_syscall_as_app() because DR needs to intercept these to
+             * interact with its multiplexing of app and client itimers (and maybe POSIX
+             * timers in the future).  We do not want to trigger our own syscall
+             * events for these, and DR indeed does not raise client events here.
+             */
+            int res =
+                dr_invoke_syscall_as_app(drcontext, SYS_timer_gettime, 2, id, &spec);
+            if (res == 0 &&
+                (!is_timespec_zero(&spec.it_interval) ||
+                 !is_timespec_zero(&spec.it_value))) {
+                if (inflate) {
+                    inflate_timespec(drcontext, &spec.it_interval);
+                    inflate_timespec(drcontext, &spec.it_value);
+                } else {
+                    deflate_timespec(drcontext, &spec.it_interval);
+                    deflate_timespec(drcontext, &spec.it_value);
+                }
+                int res = dr_invoke_syscall_as_app(drcontext, SYS_timer_settime, 4, id, 0,
+                                                   &spec, NULL);
+                if (res != 0)
+                    NOTIFY(0, "Failed to call timer_settime for id %d: %d\n", id, res);
+            } else if (res != 0) {
+                NOTIFY(0, "Failed to call timer_gettime for id %d: %d\n", id, res);
+            }
+#else
+            itimerspec64_t spec;
+            /* See above comment about dr_invoke_syscall_as_app(). */
+            int res =
+                dr_invoke_syscall_as_app(drcontext, SYS_timer_gettime64, 2, id, &spec);
+            if (res == 0 &&
+                (!is_timespec64_zero(&spec.it_interval) ||
+                 !is_timespec64_zero(&spec.it_value))) {
+                if (inflate) {
+                    inflate_timespec64(drcontext, &spec.it_interval);
+                    inflate_timespec64(drcontext, &spec.it_value);
+                } else {
+                    deflate_timespec64(drcontext, &spec.it_interval);
+                    deflate_timespec64(drcontext, &spec.it_value);
+                }
+                int res = dr_invoke_syscall_as_app(drcontext, SYS_timer_settime64, 4, id,
+                                                   0, &spec, NULL);
+                if (res != 0)
+                    NOTIFY(0, "Failed to call timer_settime64 for id %d: %d\n", id, res);
+            } else if (res != 0) {
+                NOTIFY(0, "Failed to call timer_gettime64 for id %d: %d\n", id, res);
+            }
+#endif
+        }
+    }
+    dr_close_file(fd);
+}
+
+DR_EXPORT
+bool
+drx_register_time_scaling(drx_time_scale_t *options)
+{
+    /* Only the first call succeeds. */
+    int count = dr_atomic_add32_return_sum(&init_count, 1);
+    if (count != 1) {
+        dr_atomic_add32_return_sum(&init_count, -1);
```

Comment:
I thought we usually ensure init/exit are called equal #times by keeping the init_count (unlike here where it can be only 0/1).

Reply:
See the docs which say: `* Can only be called once (until drx_unregister_time_scaling() is called).`
So we need it to fail on the 2nd call.

Changed the line 544 comment to say `    /* As documented, can only be called once (before unregister is called). */`.

Reply:
Ok so essentially we don't support multiple clients using this.